File: ch4/Problems/4.23/prime_number_generator.py
import threading
import math
import sys

class PrimeNumberGenerator(threading.Thread):
    def __init__(self, limit):
        super(PrimeNumberGenerator, self).__init__()
        self.limit = limit


    def run(self):
        print(f"Prime numbers up to {self.limit}:")
        for number in range(2, self.limit + 1):
            if self.is_prime(number):
                print(number, end=" ")
        print()


    def is_prime(self, num):
        if num <= 1:
            return False
        
        for i in range(2, int(math.sqrt(num)) + 1):
            if num % i == 0:
                return False
            
        return True
    

def main():
    if len(sys.argv) != 2:
        print("Usage: python prime_number_generator.py <number>")
        return
    
    try:
        number = int(sys.argv[1])
    except ValueError:
        print("Invalid input. Please enter a valid number.")
        return
    
    prime_thread = PrimeNumberGenerator(number)
    prime_thread.start()
    prime_thread.join()

    print("Main thread exits.")

if __name__ == "__main__":
    main()

File: ch4/Problems/4.23/PrimeNumberGenerator.java

public class PrimeNumberGenerator extends Thread {
    private int limit;

    public PrimeNumberGenerator(int limit) {
        this.limit = limit;
    }

    @Override
    public void run() {
        System.out.println("Prime numbers up to " + limit + ":");
        for (int number = 2; number <= limit; number++) {
            if (isPrime(number)) {
                System.out.print(number + "  ");
            }
        }
        System.out.println();
    }

    private boolean isPrime(int num) {
        if (num <= 1) {
            return false;
        }

        for (int i = 2; i <= Math.sqrt(num); i++) {
            if (num % i == 0) {
                return false;
            }
        }
        return true;
    }

    public static void main(String[] args) {
        if (args.length != 1) {
            System.out.println("Usage: java PrimeNumberGenerator <number>");
            return;
        }

        int number = Integer.parseInt(args[0]);

        PrimeNumberGenerator primeThread = new PrimeNumberGenerator(number);
        primeThread.start();

        try {
            primeThread.join();
        } catch (InterruptedException e) {
            e.printStackTrace();
        }

        System.out.println("Main thread exits.");
    }
}